// @angular/material/esm2020/legacy-table/table-module.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { MatLegacyRecycleRows, MatLegacyTable } from './table';
import { CdkTableModule } from '@angular/cdk/table';
import { MatLegacyCell, MatLegacyCellDef, MatLegacyColumnDef, MatLegacyFooterCell, MatLegacyFooterCellDef, MatLegacyHeaderCell, MatLegacyHeaderCellDef, } from './cell';
import { MatLegacyFooterRow, MatLegacyFooterRowDef, MatLegacyHeaderRow, MatLegacyHeaderRowDef, MatLegacyRow, MatLegacyRowDef, MatLegacyNoDataRow, } from './row';
import { MatLegacyTextColumn } from './text-column';
import { MatCommonModule } from '@angular/material/core';
import * as i0 from "@angular/core";
const EXPORTED_DECLARATIONS = [
    // Table
    MatLegacyTable,
    MatLegacyRecycleRows,
    // Template defs
    MatLegacyHeaderCellDef,
    MatLegacyHeaderRowDef,
    MatLegacyColumnDef,
    MatLegacyCellDef,
    MatLegacyRowDef,
    MatLegacyFooterCellDef,
    MatLegacyFooterRowDef,
    // Cell directives
    MatLegacyHeaderCell,
    MatLegacyCell,
    MatLegacyFooterCell,
    // Row directives
    MatLegacyHeaderRow,
    MatLegacyRow,
    MatLegacyFooterRow,
    MatLegacyNoDataRow,
    MatLegacyTextColumn,
];
/**
 * @deprecated Use `MatTableModule` from `@angular/material/table` instead. See https://material.angular.io/guide/mdc-migration for information about migrating.
 * @breaking-change 17.0.0
 */
export class MatLegacyTableModule {
}
MatLegacyTableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.0-rc.1", ngImport: i0, type: MatLegacyTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MatLegacyTableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.0-rc.1", ngImport: i0, type: MatLegacyTableModule, declarations: [
        // Table
        MatLegacyTable,
        MatLegacyRecycleRows,
        // Template defs
        MatLegacyHeaderCellDef,
        MatLegacyHeaderRowDef,
        MatLegacyColumnDef,
        MatLegacyCellDef,
        MatLegacyRowDef,
        MatLegacyFooterCellDef,
        MatLegacyFooterRowDef,
        // Cell directives
        MatLegacyHeaderCell,
        MatLegacyCell,
        MatLegacyFooterCell,
        // Row directives
        MatLegacyHeaderRow,
        MatLegacyRow,
        MatLegacyFooterRow,
        MatLegacyNoDataRow,
        MatLegacyTextColumn], imports: [CdkTableModule, MatCommonModule], exports: [MatCommonModule, 
        // Table
        MatLegacyTable,
        MatLegacyRecycleRows,
        // Template defs
        MatLegacyHeaderCellDef,
        MatLegacyHeaderRowDef,
        MatLegacyColumnDef,
        MatLegacyCellDef,
        MatLegacyRowDef,
        MatLegacyFooterCellDef,
        MatLegacyFooterRowDef,
        // Cell directives
        MatLegacyHeaderCell,
        MatLegacyCell,
        MatLegacyFooterCell,
        // Row directives
        MatLegacyHeaderRow,
        MatLegacyRow,
        MatLegacyFooterRow,
        MatLegacyNoDataRow,
        MatLegacyTextColumn] });
MatLegacyTableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.0-rc.1", ngImport: i0, type: MatLegacyTableModule, imports: [CdkTableModule, MatCommonModule, MatCommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.0-rc.1", ngImport: i0, type: MatLegacyTableModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CdkTableModule, MatCommonModule],
                    exports: [MatCommonModule, EXPORTED_DECLARATIONS],
                    declarations: EXPORTED_DECLARATIONS,
                }]
        }] });
