// @angular/core/schematics/migrations/router-link-with-href/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/router-link-with-href/util", ["require", "exports", "typescript", "@angular/core/schematics/utils/typescript/imports", "@angular/core/schematics/utils/typescript/nodes"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.migrateFile = exports.routerModule = exports.routerLinkWithHref = exports.routerLink = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    const imports_1 = require("@angular/core/schematics/utils/typescript/imports");
    const nodes_1 = require("@angular/core/schematics/utils/typescript/nodes");
    exports.routerLink = 'RouterLink';
    exports.routerLinkWithHref = 'RouterLinkWithHref';
    exports.routerModule = '@angular/router';
    function migrateFile(sourceFile, typeChecker, rewrite) {
        const routerLinkWithHrefSpec = (0, imports_1.getImportSpecifier)(sourceFile, exports.routerModule, exports.routerLinkWithHref);
        // No `RouterLinkWithHref` found, nothing to migrate, exit early.
        if (routerLinkWithHrefSpec === null)
            return;
        let rewrites = findUsages(sourceFile, typeChecker);
        // There are some usages of the `RouterLinkWithHref` symbol, which need to
        // be rewritten to `RouterLink` instead. Let's check if the `RouterLink` is
        // already imported.
        const routerLinkSpec = (0, imports_1.getImportSpecifier)(sourceFile, exports.routerModule, exports.routerLink);
        if (routerLinkSpec) {
            // The `RouterLink` symbol is already imported, just drop the `RouterLinkWithHref` one.
            const routerLinkNamedImports = routerLinkWithHrefSpec ?
                (0, nodes_1.closestNode)(routerLinkWithHrefSpec, typescript_1.default.SyntaxKind.NamedImports) :
                null;
            if (routerLinkNamedImports !== null) {
                // Given an original import like this one:
                // ```
                // import {RouterModule, RouterLinkWithHref, RouterLink} from '@angular/router';
                // ```
                // The code below removes the `RouterLinkWithHref` from the named imports section
                // (i.e. `{RouterModule, RouterLinkWithHref, RouterLink}`) and prints an updated
                // version (`{RouterModule, RouterLink}`) to a string, which is used as a
                // replacement.
                const rewrittenNamedImports = (0, imports_1.removeSymbolFromNamedImports)(routerLinkNamedImports, routerLinkWithHrefSpec);
                const printer = typescript_1.default.createPrinter();
                const replacement = printer.printNode(typescript_1.default.EmitHint.Unspecified, rewrittenNamedImports, sourceFile);
                rewrites.push({
                    startPos: routerLinkNamedImports.getStart(),
                    width: routerLinkNamedImports.getWidth(),
                    replacement: replacement,
                });
            }
        }
        else {
            // The `RouterLink` symbol is not imported, but the `RouterLinkWithHref` is imported,
            // so rewrite `RouterLinkWithHref` -> `RouterLink`.
            rewrites.push({
                startPos: routerLinkWithHrefSpec.getStart(),
                width: routerLinkWithHrefSpec.getWidth(),
                replacement: exports.routerLink,
            });
        }
        // Process rewrites last-to-first (based on start pos) to avoid offset shifts during rewrites.
        rewrites = sortByStartPosDescending(rewrites);
        for (const usage of rewrites) {
            rewrite(usage.startPos, usage.width, usage.replacement);
        }
    }
    exports.migrateFile = migrateFile;
    function findUsages(sourceFile, typeChecker) {
        const usages = [];
        const visitNode = (node) => {
            if (typescript_1.default.isImportSpecifier(node)) {
                // Skip this node and all of its children; imports are a special case.
                return;
            }
            if (typescript_1.default.isIdentifier(node)) {
                const importIdentifier = (0, imports_1.getImportOfIdentifier)(typeChecker, node);
                if ((importIdentifier === null || importIdentifier === void 0 ? void 0 : importIdentifier.importModule) === exports.routerModule &&
                    importIdentifier.name === exports.routerLinkWithHref) {
                    usages.push({
                        startPos: node.getStart(),
                        width: node.getWidth(),
                        replacement: exports.routerLink,
                    });
                }
            }
            typescript_1.default.forEachChild(node, visitNode);
        };
        typescript_1.default.forEachChild(sourceFile, visitNode);
        return usages;
    }
    /**
     * Sort all found usages based on their start positions in the source file in descending order (i.e.
     * last usage goes first on the list, etc). This is needed to avoid shifting offsets in the source
     * file (in case there are multiple usages) as we rewrite symbols.
     */
    function sortByStartPosDescending(rewrites) {
        return rewrites.sort((entityA, entityB) => entityB.startPos - entityA.startPos);
    }
});
