// @angular/core/schematics/migrations/relative-link-resolution/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/relative-link-resolution/util", ["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.migrateFile = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    const relativeLinkResolution = 'relativeLinkResolution';
    const knownConfigValues = new Set([`'legacy'`, `'corrected'`]);
    function migrateFile(sourceFile, rewriteFn) {
        let rewrites = [];
        const usages = getUsages(sourceFile);
        for (const { objectLiteral, property } of usages) {
            const replacementNode = typescript_1.default.factory.updateObjectLiteralExpression(objectLiteral, objectLiteral.properties.filter(prop => prop !== property));
            const printer = typescript_1.default.createPrinter();
            const replacementText = printer.printNode(typescript_1.default.EmitHint.Unspecified, replacementNode, sourceFile);
            rewrites.push({
                startPos: objectLiteral.getStart(),
                width: objectLiteral.getWidth(),
                replacement: replacementText,
            });
        }
        // Process rewrites last-to-first (based on start pos) to avoid offset shifts during rewrites.
        rewrites = sortByStartPosDescending(rewrites);
        for (const rewrite of rewrites) {
            rewriteFn(rewrite.startPos, rewrite.width, rewrite.replacement);
        }
    }
    exports.migrateFile = migrateFile;
    function getUsages(sourceFile) {
        const usages = [];
        const visitNode = (node) => {
            if (typescript_1.default.isObjectLiteralExpression(node)) {
                // Look for patterns like the following:
                // ```
                // { ... relativeLinkResolution: 'legacy', ... }
                // ```
                // or:
                // ```
                // { ... relativeLinkResolution: 'corrected', ... }
                // ```
                // If the value is unknown (i.e. not 'legacy' or 'corrected'),
                // do not attempt to rewrite (this might be an application-specific
                // configuration, not a part of Router).
                const property = node.properties.find(prop => typescript_1.default.isPropertyAssignment(prop) && typescript_1.default.isIdentifier(prop.name) &&
                    prop.name.text === relativeLinkResolution &&
                    knownConfigValues.has(prop.initializer.getText()));
                if (property) {
                    usages.push({ objectLiteral: node, property });
                }
            }
            typescript_1.default.forEachChild(node, visitNode);
        };
        typescript_1.default.forEachChild(sourceFile, visitNode);
        return usages;
    }
    /**
     * Sort all found usages based on their start positions in the source file in descending order (i.e.
     * last usage goes first on the list, etc). This is needed to avoid shifting offsets in the source
     * file (in case there are multiple usages) as we rewrite symbols.
     */
    function sortByStartPosDescending(rewrites) {
        return rewrites.sort((entityA, entityB) => entityB.startPos - entityA.startPos);
    }
});
