// @angular/core/schematics/migrations/path-match-type/transform.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/path-match-type/transform", ["require", "exports", "typescript", "@angular/core/schematics/utils/import_manager", "@angular/core/schematics/migrations/path-match-type/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.PathMatchTypeTransform = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    const import_manager_1 = require("@angular/core/schematics/utils/import_manager");
    const util_1 = require("@angular/core/schematics/migrations/path-match-type/util");
    class PathMatchTypeTransform {
        constructor(getUpdateRecorder) {
            this.getUpdateRecorder = getUpdateRecorder;
            this.printer = typescript_1.default.createPrinter();
            this.importManager = new import_manager_1.ImportManager(this.getUpdateRecorder, this.printer);
        }
        migrate(sourceFiles) {
            for (const sourceFile of sourceFiles) {
                const toMigrate = (0, util_1.findExpressionsToMigrate)(sourceFile, this.importManager);
                const recorder = this.getUpdateRecorder(sourceFile);
                for (const [oldNode, newNode] of toMigrate) {
                    recorder.updateNode(oldNode, newNode, sourceFile);
                }
            }
        }
        /** Records all changes that were made in the import manager. */
        recordChanges() {
            this.importManager.recordChanges();
        }
    }
    exports.PathMatchTypeTransform = PathMatchTypeTransform;
});
