// @angular/core/schematics/migrations/path-match-type/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/path-match-type/util", ["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.findExpressionsToMigrate = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    function findExpressionsToMigrate(sourceFile, importManager) {
        const migratedNodesMap = new Map();
        let _currentVariableDecl = null;
        (() => {
            sourceFile.forEachChild(function visitNode(node) {
                if (typescript_1.default.isVariableDeclaration(node)) {
                    _currentVariableDecl = node;
                    node.forEachChild(visitNode);
                    _currentVariableDecl = null;
                }
                if (isRouteOrRoutesVariableDeclaration(node)) {
                    // The variable declaration is already explicitly typed as `Route` or `Routes` so it does
                    // not need a migration.
                    return;
                }
                else if (typescript_1.default.isObjectLiteralExpression(node)) {
                    if (_currentVariableDecl !== null && _currentVariableDecl.type === undefined) {
                        visitObjectLiteral(node);
                    }
                }
                else {
                    node.forEachChild(visitNode);
                }
            });
            function visitObjectLiteral(obj) {
                const hasPathMatch = obj.properties.some(p => isPropertyWithName(p, 'pathMatch'));
                const hasPath = obj.properties.some(p => isPropertyWithName(p, 'path'));
                const childrenProperty = obj.properties.find(p => isPropertyWithName(p, 'children'));
                // The object must have _both_ pathMatch _and_ path for us to be reasonably sure that it's
                // a `Route` definition.
                if (hasPath && hasPathMatch) {
                    updateCurrentVariableDeclaration();
                }
                else if (childrenProperty !== undefined && typescript_1.default.isPropertyAssignment(childrenProperty) &&
                    typescript_1.default.isArrayLiteralExpression(childrenProperty.initializer)) {
                    // Also need to check the children if it exists
                    for (const child of childrenProperty.initializer.elements) {
                        if (typescript_1.default.isObjectLiteralExpression(child)) {
                            visitObjectLiteral(child);
                            // If the child caused a migration, we can exit early
                            if (_currentVariableDecl && migratedNodesMap.has(_currentVariableDecl)) {
                                break;
                            }
                        }
                    }
                }
            }
            function isPropertyWithName(p, name) {
                if (typescript_1.default.isPropertyAssignment(p)) {
                    return p.name.getText() === name;
                }
                else if (typescript_1.default.isShorthandPropertyAssignment(p)) {
                    return p.name.getText() === name;
                }
                else {
                    // Don't attempt to migrate edge case spreadAssignment
                    return false;
                }
            }
            function updateCurrentVariableDeclaration() {
                if (_currentVariableDecl === null || _currentVariableDecl.initializer === undefined) {
                    return;
                }
                let typeToUse;
                if (typescript_1.default.isArrayLiteralExpression(_currentVariableDecl.initializer)) {
                    typeToUse = importManager.addImportToSourceFile(sourceFile, 'Routes', '@angular/router');
                }
                else {
                    typeToUse = importManager.addImportToSourceFile(sourceFile, 'Route', '@angular/router');
                }
                const migrated = typescript_1.default.factory.updateVariableDeclaration(_currentVariableDecl, _currentVariableDecl.name, _currentVariableDecl.exclamationToken, typeToUse, _currentVariableDecl.initializer);
                migratedNodesMap.set(_currentVariableDecl, migrated);
            }
        })();
        return migratedNodesMap;
    }
    exports.findExpressionsToMigrate = findExpressionsToMigrate;
    function isRouteOrRoutesVariableDeclaration(node) {
        return typescript_1.default.isVariableDeclaration(node) && node.type &&
            (node.type.getText() === 'Route' || node.type.getText() === 'Routes');
    }
});
