// @angular/core/schematics/migrations/typed-forms/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/typed-forms/util", ["require", "exports", "typescript", "@angular/core/schematics/utils/typescript/imports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.migrateFile = exports.forms = exports.untypedPrefix = exports.formControl = exports.classes = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    const imports_1 = require("@angular/core/schematics/utils/typescript/imports");
    exports.classes = new Set(['FormArray', 'FormBuilder', 'FormControl', 'FormGroup']);
    exports.formControl = 'FormControl';
    exports.untypedPrefix = 'Untyped';
    exports.forms = '@angular/forms';
    function migrateFile(sourceFile, typeChecker, rewrite) {
        const imports = getImports(sourceFile);
        // If no relevant classes are imported, we can exit early.
        if (imports.length === 0)
            return;
        // For each control class, migrate all of its uses.
        for (let i = imports.length; i >= 0; i--) {
            const imp = imports[i];
            const usages = getUsages(sourceFile, typeChecker, imp);
            if (usages.length === 0) {
                // Since there are no usages of this class we need to migrate it, we should completely
                // skip it for the subsequent migration steps.
                imports.splice(i, 1);
            }
            for (const usage of usages) {
                const newName = getUntypedVersionOfImportOrName(usage.importName);
                if (newName === null) {
                    // This should never happen.
                    console.error(`Typed forms migration error: unknown replacement for usage ${usage.node.getText()}`);
                    continue;
                }
                rewrite(usage.node.getStart(), usage.node.getWidth(), newName);
            }
        }
        // For each imported control class, migrate to the corresponding uptyped import.
        for (const imp of imports) {
            const untypedClass = getUntypedVersionOfImportOrName(imp.getText());
            if (untypedClass === null) {
                // This should never happen.
                console.error(`Typed forms migration error: unknown untyped version of import ${imp.getText()}`);
                continue;
            }
            if ((0, imports_1.getImportSpecifier)(sourceFile, exports.forms, untypedClass)) {
                // In order to make the migration idempotent, we must check whether the untyped version of the
                // class is already present. If present, immediately continue.
                continue;
            }
            rewrite(imp.getStart(), imp.getWidth(), untypedClass);
        }
    }
    exports.migrateFile = migrateFile;
    function getImports(sourceFile) {
        let imports = [];
        for (const cc of exports.classes) {
            const specifier = (0, imports_1.getImportSpecifier)(sourceFile, exports.forms, cc);
            if (!specifier)
                continue;
            imports.push(specifier);
        }
        return imports;
    }
    function getUntypedVersionOfImportOrName(name) {
        for (const cc of exports.classes) {
            if (name.includes(cc)) {
                return `${exports.untypedPrefix}${cc}`;
            }
        }
        return null;
    }
    function getUsages(sourceFile, typeChecker, importSpecifier) {
        const usages = [];
        const visitNode = (node) => {
            if (typescript_1.default.isImportSpecifier(node)) {
                // Skip this node and all of its children; imports are a special case.
                return;
            }
            if (typescript_1.default.isIdentifier(node) && isUsageOfFormsImport(typeChecker, node, importSpecifier)) {
                usages.push({ node, importName: importSpecifier.getText() });
            }
            typescript_1.default.forEachChild(node, visitNode);
        };
        typescript_1.default.forEachChild(sourceFile, visitNode);
        return usages;
    }
    function isUsageOfFormsImport(typeChecker, node, importSpecifier) {
        var _a, _b;
        const symbol = typeChecker.getSymbolAtLocation(node);
        // We check symbol.declarations because we actually care about the name at the declaration site,
        // not the usage site. These could be different in the case of overridden constructors.
        if (!symbol || symbol.declarations === undefined || !symbol.declarations.length)
            return false;
        const decl = symbol.declarations[0];
        if (!typescript_1.default.isImportSpecifier(decl))
            return false;
        // As per `typescript/imports.ts`, we must walk up the tree to find the enclosing import
        // declaration. For reasons specific to the TS AST, this is always 3 levels up from an import
        // specifier node.
        const importDecl = decl.parent.parent.parent;
        if (!typescript_1.default.isStringLiteral(importDecl.moduleSpecifier))
            return false;
        const importName = (_b = (_a = typeChecker.getTypeAtLocation(importSpecifier)) === null || _a === void 0 ? void 0 : _a.getSymbol()) === null || _b === void 0 ? void 0 : _b.getName();
        if (!importName)
            return false;
        // Handles aliased imports: e.g. "import {Component as myComp} from ...";
        const declName = decl.propertyName ? decl.propertyName.text : decl.name.text;
        if (importName === declName)
            return true;
        // In the case of FormControl's overridden exported constructor, the value name and declaration
        // name are not exactly the same. For our purposes, it's enough to check whether the latter is a
        // substring of the former.
        if (declName === exports.formControl && importName.includes(declName))
            return true;
        return false;
    }
});
