// @angular/material/esm2020/tabs/public-api.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export * from './module';
export { MatPaginatedTabHeader } from './paginated-tab-header';
export { MatTabBodyPortal, MatTabBody, _MatTabBodyBase, } from './tab-body';
export { MAT_TABS_CONFIG } from './tab-config';
export { MatTabContent, MAT_TAB_CONTENT } from './tab-content';
export { MatTabLabel, MAT_TAB, MAT_TAB_LABEL } from './tab-label';
export { MatTabLabelWrapper } from './tab-label-wrapper';
export { MatTab, MAT_TAB_GROUP, _MatTabBase } from './tab';
export { MatInkBar, _MAT_INK_BAR_POSITIONER_FACTORY, _MAT_INK_BAR_POSITIONER, } from './ink-bar';
export { MatTabHeader, _MatTabHeaderBase } from './tab-header';
export { MatTabGroup, MatTabChangeEvent, _MatTabGroupBase, } from './tab-group';
export { MatTabNav, MatTabNavPanel, MatTabLink, _MatTabNavBase, _MatTabLinkBase, } from './tab-nav-bar/tab-nav-bar';
export { matTabsAnimations } from './tabs-animations';
export { _MatTabLabelWrapperBase } from './tab-label-wrapper';
